src/core/thread/network_data_notifier.cpp

```diff
@@ -0,0 +1,181 @@
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements transmissions of SVR_DATA.ntf messages to the Leader.
+ */
+
+#include "network_data_notifier.hpp"
+
+#if OPENTHREAD_FTD || OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE || OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE
+
+#include "common/code_utils.hpp"
+#include "common/instance.hpp"
+#include "common/locator-getters.hpp"
+#include "thread/network_data_leader.hpp"
+#include "thread/network_data_local.hpp"
+
+namespace ot {
+namespace NetworkData {
+
+Notifier::Notifier(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mNotifierCallback(aInstance, &Notifier::HandleStateChanged, this)
+    , mTimer(aInstance, &Notifier::HandleTimer, this)
+    , mState(kStateNewServiceData)
+    , mWaitingForResponse(false)
+{
+}
+
+void Notifier::ServerDataUpdated(void)
+{
+    mState = kStateNewServiceData;
+    SynchronizeNetworkData();
+}
+
+void Notifier::SynchronizeNetworkData(void)
+{
+    otError error = OT_ERROR_NOT_FOUND;
+
+    VerifyOrExit(Get<Mle::MleRouter>().IsAttached() && !mWaitingForResponse);
+
+    VerifyOrExit((mState == kStateNewServiceData) || !mTimer.IsRunning());
+
+#if OPENTHREAD_FTD
+    mState = kStateRemoveStaleChildren;
+    error  = Get<Leader>().RemoveStaleChildEntries(&Notifier::HandleCoapResponse, this);
+    VerifyOrExit(error == OT_ERROR_NOT_FOUND);
+#endif
+
+#if OPENTHREAD_CONFIG_BORDER_ROUTER_ENABLE || OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE
+    mState = kStateSynchronizeServerData;
+    error  = Get<Local>().UpdateInconsistentServerData(&Notifier::HandleCoapResponse, this);
+    VerifyOrExit(error == OT_ERROR_NOT_FOUND);
+#endif
+
+    mState = kStateSynchronized;
+
+exit:
+    switch (error)
+    {
+    case OT_ERROR_NONE:
+        mWaitingForResponse = true;
+        break;
+    case OT_ERROR_NO_BUFS:
+    case OT_ERROR_INVALID_STATE:
```

Comment:
`OT_ERROR_INVALID_STATE` error happens when the server is in the routerupgradeselection jitter period, the 1s timer (`kDelayNoBufs`) may be repeatedly triggered due to the same error in this period.
Different from `OT_ERROR_NO_BUFS`,  we could know the time required to exit `OT_ERROR_INVALID_STATE` via  [GetRouterSelectionJitterTimeout()](https://github.com/openthread/openthread/blob/master/src/core/thread/mle_router.hpp#L288)?  how about starting a timer based on that value?